### DISCUSSION TITLE
How to dummy ButtonPress and ButtonRelease events in X11 client?

### DISCUSSION THREAD

**Discussion by Saddingt0n** [FreeRDP/FreeRDP#8488] [Q&A] [answered]
My apology if the title is not phrased correctly.

I'm modifying X11 client and I want to do something like this. The client is to take input from the keyboard in the format of `"<event type> <x coordinate> <y coordinate>"`.
Then it will perform a mouse action accordingly. For example: "press 50 50" will do a mouse press at (50, 50). 

I dig around the source code and found that the `xf_event_process` function in `client/X11/xf_event.c` handles various events, including `ButtonPress` and `ButtonRelease`.
However, this function takes a `freerdp` pointer and a `XEvent` pointer as arguments. I don't know how to get these 2 as arguments. I found out that the `x` and `y` fields in `XMotionEvent` field of `XEvent` contains the absolute coordinates of cursor on screen:
![image](https://user-images.githubusercontent.com/91487625/205609641-c8cb8f50-1049-46ec-ada5-dce931b9b8bf.png)

Could someone give me some pointers to do what I have described? I know it's a bit unorthodox.

**Comment by akallabeth** [comment, marked as answer]
1. will not work that easily as `xfreerdp` captures the keyboard input too. (you´d have to disable keyboard support in `xfreerdp` for your thing to work)
2. Faking input you can do externally with [xdotool](https://www.semicomplete.com/projects/xdotool/)

**Reply by Saddingt0n** [reply]
> 1. will not work that easily as `xfreerdp` captures the keyboard input too. (you´d have to disable keyboard support in `xfreerdp` for your thing to work)
> 2. Faking input you can do externally with [xdotool](https://www.semicomplete.com/projects/xdotool/)

Thank you, @akallabeth 

1. I'm not sure what you mean. Could you elaborate further?
2. I have tried xdotool tool but it seems that it can only do input when the target window is on screen. I want to be able to do it when the window is not on screen/minized. Could you recommend another app or workaround?

**Reply by akallabeth** [reply]
1. `xfreerdp` captures keyboard events and redirects them to the remote end. aside from simple keys you can not assign combinations easily without breaking something remote.
2. `xdotool --window <id>` should do the trick, does it not?

**Reply by Saddingt0n** [reply]
> 1. `xfreerdp` captures keyboard events and redirects them to the remote end. aside from simple keys you can not assign combinations easily without breaking something remote.
> 2. `xdotool --window <id>` should do the trick, does it not?

1. Thank you but I was asking about mouse click. Do mouse click events work similarly to keyboard events? I want to perform simple left click, release.
2. It works but only when FreeRDP window is visible on screen. Thank you though, you've been very helpful.
